src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -259,33 +264,44 @@ internal Task<Document> GetChangedDocumentInternalAsync(CancellationToken cancel
         /// <param name="operations">A list of operations.</param>
         /// <param name="cancellationToken">A cancellation token.</param>
         /// <returns>A new list of operations with post processing steps applied to any <see cref="ApplyChangesOperation"/>'s.</returns>
-        protected async Task<ImmutableArray<CodeActionOperation>> PostProcessAsync(IEnumerable<CodeActionOperation> operations, CancellationToken cancellationToken)
+        protected Task<ImmutableArray<CodeActionOperation>> PostProcessAsync(IEnumerable<CodeActionOperation> operations, CancellationToken cancellationToken)
+            => PostProcessAsync(originalSolution: null!, operations, cancellationToken);
+
+        internal async Task<ImmutableArray<CodeActionOperation>> PostProcessAsync(
+            Solution originalSolution, IEnumerable<CodeActionOperation> operations, CancellationToken cancellationToken)
         {
-            var arrayBuilder = new ArrayBuilder<CodeActionOperation>();
+            using var result = TemporaryArray<CodeActionOperation>.Empty;
 
             foreach (var op in operations)
             {
                 if (op is ApplyChangesOperation ac)
                 {
-                    arrayBuilder.Add(new ApplyChangesOperation(await this.PostProcessChangesAsync(ac.ChangedSolution, cancellationToken).ConfigureAwait(false)));
+                    result.Add(new ApplyChangesOperation(await this.PostProcessChangesAsync(originalSolution, ac.ChangedSolution, cancellationToken).ConfigureAwait(false)));
                 }
                 else
                 {
-                    arrayBuilder.Add(op);
+                    result.Add(op);
                 }
             }
 
-            return arrayBuilder.ToImmutableAndFree();
+            return result.ToImmutableAndClear();
         }
 
         /// <summary>
-        ///  Apply post processing steps to solution changes, like formatting and simplification.
+        /// Apply post processing steps to solution changes, like formatting and simplification.
         /// </summary>
         /// <param name="changedSolution">The solution changed by the <see cref="CodeAction"/>.</param>
         /// <param name="cancellationToken">A cancellation token</param>
-        protected async Task<Solution> PostProcessChangesAsync(Solution changedSolution, CancellationToken cancellationToken)
+        protected Task<Solution> PostProcessChangesAsync(Solution changedSolution, CancellationToken cancellationToken)
+            => PostProcessChangesAsync(originalSolution: null!, changedSolution, cancellationToken);
+
+        internal async Task<Solution> PostProcessChangesAsync(
+            Solution originalSolution,
+            Solution changedSolution,
+            CancellationToken cancellationToken)
         {
-            var solutionChanges = changedSolution.GetChanges(changedSolution.Workspace.CurrentSolution);
+            originalSolution ??= changedSolution.Workspace.CurrentSolution;
```

Comment:
>originalSolution [](http://example.com/codeflow?start=12&length=16)

I think this warrants a comment that originalSolution is null only on backward compat code paths.

Reply:
definitely.  will add.

Reply:
done.